components/provider_cldr/src/transform/timezones/cldr_json/mod.rs

```diff
@@ -0,0 +1,167 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! Data provider struct definitinos for mapping CLDR-JSON Time Zones.
+//! These provide in-memory stuctures into which `timeZoneNames.json` will be serialized.
+//! Here is the `en` [timeZoneNames.json](https://raw.githubusercontent.com/unicode-org/cldr-json/master/cldr-json/cldr-dates-full/main/en/timeZoneNames.json) for context.
+
+mod convert;
+
+use crate::cldr_langid::CldrLangID;
+use serde::{
+    de::{IgnoredAny, MapAccess, Visitor},
+    Deserialize, Deserializer,
+};
+use std::collections::BTreeMap;
+
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub struct ZoneFormat(pub BTreeMap<String, String>);
+
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub struct MetaZone {
+    pub long: Option<ZoneFormat>,
+    pub short: Option<ZoneFormat>,
+}
+
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub struct MetaZones(pub BTreeMap<String, MetaZone>);
+
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub struct LocationWithExemplarCity {
+    pub long: Option<ZoneFormat>,
+    pub short: Option<ZoneFormat>,
+    #[serde(rename = "exemplarCity")]
+    pub exemplar_city: String,
+}
+
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub struct LocationWithShort {
+    pub long: Option<ZoneFormat>,
+    pub short: ZoneFormat,
+    #[serde(rename = "exemplarCity")]
+    pub exemplar_city: Option<String>,
+}
+
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+pub struct LocationWithLong {
+    pub long: ZoneFormat,
+    pub short: Option<ZoneFormat>,
+    #[serde(rename = "exemplarCity")]
+    pub exemplar_city: Option<String>,
+}
+
+#[derive(PartialEq, Debug, Clone, Deserialize)]
+#[serde(untagged)]
+pub enum Location {
+    LocationE(LocationWithExemplarCity),
+    LocationL(LocationWithLong),
+    LocationS(LocationWithShort),
```

Comment:
`suggestion`: if there is no strong reason, I'd vastly prefer `LocationCity`, `LocationLong`, `LocationShort` over that. Reading code and having to skip around to derive the meaning of `LocationL` imposes a maintenance tax.

Reply:
Ah yeah, I think you're right. 

I was avoiding calling it `LocationLong` or `LocationShort` because it's not a formatting thing. It's just which one is required in the particular variant. 

Having all three variants was an unfortunate necessity in order to get Serde to parse the otherwise ambiguous JSON, using each variant as a fallback if the previous one failed to parse. All three must be exhausted before we can be sure that it's a nested region (with three segments instead of two), such as  `America/Argentina/Rio_Gallegos`.

I will think of names that are more descriptive than a single letter.